fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/ValuePickerValue.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * QueriedValue.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.predicates;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.annotation.SpotBugsSuppressWarnings;
+import com.apple.foundationdb.record.ObjectPlanHash;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.query.plan.temp.AliasMap;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * A value representing multiple "alternative" values.
+ * This value holds multiple "alternatives" for values and allows the creator to select one of them to be returned as
+ * the "actual" value.
+ * This is useful in cases (like {@link com.apple.foundationdb.record.query.plan.plans.RecordQuerySelectorPlan}) where
+ * there can be one of several plans that can be selected for execution, and this values gives visibility to all
+ * of their {@link Value}s
+ */
+@API(API.Status.EXPERIMENTAL)
+public class ValuePickerValue implements Value.CompileTimeValue {
```

Comment:
Why? That is, how does deferring the selection of the value to use until such use help? Why not just remember the chosen value at the place that would otherwise call one of these constructors?

Reply:
This is per the design by @normen662 (similar to the type system design) where the entire `Value` population for the plan is kept around for the cases where it is needed. This way, the actual executed plan makes its runtime `values` available, but exploration of the possible values is still available.